src/c1_foeppl_von_karman/foeppl_von_karman_curvable_bell_elements.h

```diff
@@ -1175,1704 +1561,877 @@ namespace oomph
+  TriangleMeshCurviLine* curviline_pt)
+ {
+
+  for (unsigned i=0;i<3;i++)
+   {
+    pin_and_impose_specified_displacement_along_specified_boundary
+     (i,b,boundary_values_pt[i],curviline_pt);
+   }
+
+ }
+ 
+
+ //=============================================================================
+ /// Helper function to impose alpha-th in-plane displacements
+ ///(0 or 1 for x or y displacements) according to scalar function
+ /// specified in boundary_values_pt. curvline_pt provides the
+ /// geometry of the boundary in terms of a not-necessarily-arclength
+ /// coordinate zeta.
+ //=============================================================================
+ template<unsigned NNODE_1D>
+ void FoepplVonKarmanC1CurvableBellElement<NNODE_1D>::
+ pin_and_impose_specified_in_plane_displacement_along_specified_boundary(
+  const unsigned& alpha,
+  const unsigned& b,
+  BoundaryConditionForC1PlateBending* boundary_values_pt,
+  TriangleMeshCurviLine* curviline_pt)
+ {
+    
+  // Initialise constants that we use in this function
+  const unsigned field_index = u_alpha_field_index(alpha);
+  const unsigned nodal_type_index =
+   this->first_nodal_type_index_for_field(field_index);
+  const unsigned n_node = nu_node();
+  const unsigned dim = this->dim();
+  
+#ifdef PARANOID
+  // Check that the dof number is a sensible value
+  const unsigned n_type = 2 * nu_type_at_each_node();
+  if (alpha >= n_type)
+   {
+    throw OomphLibError(
+     "Foppl von Karman elements only have 2 in-plane displacement degrees\
+of freedom at internal points. They are {ux, uy}",
+     OOMPH_CURRENT_FUNCTION,
+     OOMPH_EXCEPTION_LOCATION);
+   }
+#endif
+  
+  // Bell elements only have deflection dofs at vertices
+  for (unsigned n = 0; n < n_node; ++n)
+   {
+    // Get boundary node
+    BoundaryNode<Node>* nod_pt =
+     dynamic_cast<BoundaryNode<Node>*>(this->node_pt(n));
+    
+    // Check if it is on the boundary
+    if (nod_pt!=0)
+     {
+      if (nod_pt->is_on_boundary(b))
+       {
+#ifdef PARANOID
+        // We should only have one coordinate on this boundary
+        unsigned nzeta=nod_pt->ncoordinates_on_boundary(b);
+        if (nzeta!=1)
+         {
+          std::stringstream error_message;
+          error_message << "boundary coordinate must be 1D!"
+                        << std::endl;
+          throw OomphLibError(error_message.str(),
+                              OOMPH_CURRENT_FUNCTION,
+                              OOMPH_EXCEPTION_LOCATION);
+         }
+#endif
+        
+        // Get value itself from boundary condition object
+        Vector<double> zeta(nzeta);
+        nod_pt->get_coordinates_on_boundary(b,zeta);
+        double value=boundary_values_pt->f(zeta[0]);
+                
+        // Pin and set the value
+        nod_pt->pin(alpha);
+        nod_pt->set_value(nodal_type_index, value); // hierher Aidan should indices in pin and set_value be the same? 
```

Comment:
Yes, both should be nodal_type_index. This function has been made overly generic in case non-Lagrange interpolation is ever used for the in-plane dofs.